### PR TITLE
[Storage] return ValidatorChangeEventWithProof in LibraDB::update_to_latest_ledger

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

As requested by consensus side. Also refactored LibraDB::get_latest_ledger_info_per_epoch to skip returning the latest ledger info if it's not an epoch bump, because on both call sites of it we filter that out.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

unit tests

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: storage/libradb/src/ledger_store/mod.rs [hunk 1]
```diff
@@ -41,6 +41,8 @@ pub(crate) struct LedgerStore {
     latest_ledger_info: ArcSwap<Option<LedgerInfoWithSignatures>>,
 }
 
+const MAX_NUM_EPOCH_CHANGE_LEDGER_INFO: usize = 100;
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]: I think this might not make sense, at least right now. If a client has a really old `client_known_version` such that there are more than `MAX_NUM_EPOCH_CHANGE_LEDGER_INFO` epoch changes between `client_known_version` and the latest version, the implementation below will always return an error and there's no way for the client to catch up. Do I understand this correctly?

File: storage/libradb/src/lib.rs [hunk 2]
```diff
@@ -538,8 +539,9 @@ impl LibraDB {
             })
             .collect::<Result<Vec<_>>>()?;
 
+        // TODO: Cache last epoch change version, to avoid a DB access in most cases.
         let client_epoch = self.ledger_store.get_epoch(client_known_version)?;
-        let validator_change_proof = if client_epoch < ledger_info_with_sigs.ledger_info().epoch() {
+        let validator_change_proof = if client_epoch < self.ledger_store.get_current_epoch()? {
```
Comment (wqfish) [inline storage/libradb/src/lib.rs]: Why this change? Doesn't this lead to some race condition, since we now call `get_latest_ledger_info` twice in this function and they may return different version if there's a concurrent write?

File: storage/libradb/src/ledger_store/mod.rs (hunk 1 above)
Comment (msmouse) [reply storage/libradb/src/ledger_store/mod.rs]: Yes we need to implement a protocol to hint the client to loop or boot from a waypoint. This protects the server only.

File: storage/libradb/src/lib.rs (hunk 2 above)
Comment (msmouse) [reply storage/libradb/src/lib.rs]: This changed because of the case that the latest LedgerInfo has Some(next_validator_set) hence the real current epoch number is latest_LI.epoch() + 1.

But it's a good point that the latest ledger_info can change when the request is running. We need to take that into consideration (not only this if-condition, but also skip returning info about epochs newer than the current epoch indicated by `ledger_version`.

File: storage/libradb/src/ledger_store/mod.rs (hunk 1 above)
Comment (wqfish) [reply storage/libradb/src/ledger_store/mod.rs]: I'd just document this a little bit in the code so anyone who reads the code is able to understand.

File: storage/libradb/src/ledger_store/mod.rs [hunk 3]
```diff
@@ -91,7 +93,17 @@ impl LedgerStore {
     ) -> Result<Vec<LedgerInfoWithSignatures>> {
         let mut iter = self.db.iter::<LedgerInfoSchema>(ReadOptions::default())?;
         iter.seek(&start_epoch)?;
-        Ok(iter.map(|kv| Ok(kv?.1)).collect::<Result<Vec<_>>>()?)
+        let iter = iter.take(MAX_NUM_EPOCH_CHANGE_LEDGER_INFO + 1);
+        let result = iter.map(|res| Ok(res?.1)).collect::<Result<Vec<_>>>()?;
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]:
```
let result = iter.take(...).map(...).collect(...)?;
```

File: storage/libradb/src/lib.rs [hunk 4]
```diff
@@ -538,14 +539,39 @@ impl LibraDB {
...
+            if ledger_infos
+                .last()
+                .ok_or_else(|| LibraDbError::NotFound(String::from("Latest epoch LedgerInfo")))?
+                .ledger_info()
+                .next_validator_set()
+                .is_none()
+            {
+                ledger_infos.pop();
+            }
+            ledger_infos
```
Comment (wqfish) [inline storage/libradb/src/lib.rs]: Does this have the same problem? Imagine there is an epoch change that happens right after the `get_latest_ledger_info` call at line 466 and before this piece of code is executed. Then this might return the newly added epoch change, which is not desired?

Comment (wqfish) [reply storage/libradb/src/lib.rs]: Looks like a later commit fixes this?

File: storage/libradb/src/ledger_store/mod.rs [hunk 5]
```diff
@@ -82,18 +82,31 @@ impl LedgerStore {
...
+        let iter = iter
+            .take(MAX_NUM_EPOCH_CHANGE_LEDGER_INFO + 1)
+            .take_while(|res| {
+                res.as_ref()
+                    .ok()
+                    .map(|(_, x)| {
+                        x.ledger_info().version() <= ledger_version
+                            && x.ledger_info().next_validator_set().is_some()
+                    })
+                    .unwrap_or(false)
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]: If `res` is an error, this would cause the scan to stop early and the code below to return a shorter result, instead of returning an error.

File: storage/libradb/src/lib.rs [hunk 6]
```diff
@@ -304,18 +304,17 @@ impl LibraDB {
...
     }
 
-    /// Returns the latest ledger infos per epoch starting with the given epoch num:
-    /// - the latest ledger info of the current epoch is just the last ledger info in the system
-    /// - the latest ledger infos of previous epochs contain reconfiguration validator sets.
+    /// Returns ledger infos reflecting epoch bumps starting with the given epoch.
+    ///
     /// Returns error in case `start_epoch` is higher than the currently known epoch.
     /// The returned vector is not necessarily sorted: the client should make sure to sort it
     /// by epoch number.
```
Comment (wqfish) [inline storage/libradb/src/lib.rs]: This looks weird.

File: storage/libradb/src/lib.rs [hunk 7]
```diff
@@ -304,18 +304,17 @@ impl LibraDB {
             .version())
     }
 
-    /// Returns the latest ledger infos per epoch starting with the given epoch num:
-    /// - the latest ledger info of the current epoch is just the last ledger info in the system
-    /// - the latest ledger infos of previous epochs contain reconfiguration validator sets.
+    /// Returns ledger infos reflecting epoch bumps starting with the given epoch.
+    ///
     /// Returns error in case `start_epoch` is higher than the currently known epoch.
```
Comment (wqfish) [inline storage/libradb/src/lib.rs]: This is probably not true.

File: storage/libradb/src/lib.rs (hunk 4 above)
Comment (msmouse) [reply storage/libradb/src/lib.rs]: Yes the third commit resolves this.

File: storage/libradb/src/ledger_store/mod.rs (hunk 5 above)
Comment (msmouse) [reply storage/libradb/src/ledger_store/mod.rs]: Gooooood catch!

File: storage/libradb/src/lib.rs (hunk 7 above)
Comment (msmouse) [reply storage/libradb/src/lib.rs]: yeah.. but I didn't touch it.. probably removing..

File: storage/libradb/src/lib.rs (hunk 6 above)
Comment (msmouse) [reply storage/libradb/src/lib.rs]: ditto

File: storage/libradb/src/ledger_store/mod.rs [hunk 8]
```diff
@@ -41,6 +41,10 @@ pub(crate) struct LedgerStore {
     latest_ledger_info: ArcSwap<Option<LedgerInfoWithSignatures>>,
 }
 
+// TODO: Either implement an iteration API to allow a very old client to loop through a long history
+// or guarantee that there is always a recent enough waypoint and client knowns to boot from there.
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]:
```suggestion
// or guarantee that there is always a recent enough waypoint that client knows to boot from there.
```

File: storage/libradb/src/ledger_store/mod.rs [hunk 9]
```diff
@@ -41,6 +41,8 @@ pub(crate) struct LedgerStore {
...
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]: **knows**?

File: storage/libradb/src/ledger_store/mod.rs [hunk 10]
```diff
@@ -93,19 +95,22 @@ impl LedgerStore {
...
+                        // the epoch.
                         x.ledger_info().version() <= ledger_version
                             && x.ledger_info().next_validator_set().is_some()
                     })
-                    .unwrap_or(false)
-            });
-
-        let result = iter.map(|res| Ok(res?.1)).collect::<Result<Vec<_>>>()?;
+                    // Pass through Err.
+                    .unwrap_or(true)
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]: The whole thing is still not very easy to read. Maybe writing a loop would be better.

File: storage/libradb/src/ledger_store/mod.rs [hunk 11]
```diff
@@ -42,7 +42,7 @@ pub(crate) struct LedgerStore {
 }
 
 // TODO: Either implement an iteration API to allow a very old client to loop through a long history
-// or guarantee that there is always a recent enough waypoint and client knowns to boot from there.
+// or guarantee that there is always a recent enough waypoint and client knows to boot from there.
```
Comment (msmouse) [inline storage/libradb/src/ledger_store/mod.rs]: @zekun000 I'm sure we have a plan to deal with this situation, right?

Comment (zekun000) [reply storage/libradb/src/ledger_store/mod.rs]: I don't have a plan for client yet, but for validators it'll keep requesting epoch change proof if it receives future epoch messages until it catches up

Comment (zekun000) [reply storage/libradb/src/ledger_store/mod.rs]: hmm, I realize the behavior is not returning at most 100 items but return error if it exceeds 100, and there's no API for client to pass the limit

seems there's no way to loop through from very far in this case?

Comment (lightmark) [reply storage/libradb/src/ledger_store/mod.rs]: Can we set a limit with a tag like EOF telling user whether we reach the latest?

Comment (msmouse) [reply storage/libradb/src/ledger_store/mod.rs]: @zekun000 Yes, so the plan is the "looping" instead of "guaranteeing there is a recent enough waypoint"? We need to implement the API with cursor.

@lightmark a user can compare the epoch in the last epoch change li with the one on the latest li to know if it has reached the latest.
